runtime/commonMain/src/kotlinx/serialization/internal/SerialClassDescImpl.kt

```diff
@@ -34,13 +39,24 @@ open class SerialClassDescImpl @JvmOverloads constructor(
         val idx = names.size - 1
         ensureFlagsCapacity(idx)
         flags[idx] = isOptional
-        annotations.add(mutableListOf())
+        propertiesAnnotations.add(null)
     }
 
-    public fun pushAnnotation(a: Annotation) {
-        annotations.last().add(a)
+    // TODO rename
+    public fun pushAnnotation(annotation: Annotation) {
+        val list = propertiesAnnotations.last().let {
+            if (it == null) {
+                val result = ArrayList<Annotation>(1)
+                propertiesAnnotations[propertiesAnnotations.lastIndex] = result
+                result
+            } else {
+                it
+            }
+        }
+        list.add(annotation)
     }
 
+    // TODO rename
     public fun pushClassAnnotation(a: Annotation) {
```

Comment:
Probably annotate these methods with `@InternalSerializationApi`? I feel that when the type is not explicitly written there will be no warning in IDE

Reply:
Not yet. Will work on changing this class API and internals in a separate PR

Reply:
(same for other TODOs)